nomos-services/mix/src/backends/libp2p.rs

```diff
@@ -0,0 +1,224 @@
+            .with_behaviour(|_| {
+                nomos_mix_network::Behaviour::new(nomos_mix_network::Config {
+                    transmission_rate: 1.0,
+                    duplicate_cache_lifespan: 60,
+                })
+            })
+            .expect("Mix Behaviour should be built")
+            .with_swarm_config(|cfg| {
+                cfg.with_idle_connection_timeout(Duration::from_secs(u64::MAX))
+            })
+            .build();
+
+        Self {
+            swarm,
+            num_mix_layers,
+            msgs_rx,
+            events_tx,
+        }
+    }
+
+    fn listen_on(&mut self, addr: Multiaddr) -> Result<ListenerId, TransportError<io::Error>> {
+        self.swarm.listen_on(addr)
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<(), DialError> {
+        self.swarm.dial(DialOpts::from(addr))
+    }
+
+    async fn run(&mut self) {
+        loop {
+            tokio::select! {
+                Some(msg) = self.msgs_rx.recv() => {
+                    self.handle_msg(msg).await;
+                }
+                Some(event) = self.swarm.next() => {
+                    self.handle_event(event);
+                }
+            }
+        }
+    }
+
+    async fn handle_msg(&mut self, msg: Libp2pNetworkBackendMessage) {
+        match msg {
+            Libp2pNetworkBackendMessage::Mix(msg) => {
+                tracing::debug!("Wrap msg and send it to mix network: {msg:?}");
+                match nomos_mix_message::new_message(&msg, self.num_mix_layers.try_into().unwrap())
+                {
+                    Ok(wrapped_msg) => {
+                        if let Err(e) = self.swarm.behaviour_mut().publish(wrapped_msg) {
+                            tracing::error!("Failed to publish message to mix network: {e:?}");
+                        }
+                    }
+                    Err(e) => {
+                        tracing::error!("Failed to wrap message: {e:?}");
+                    }
+                }
+            }
+        }
+    }
+
+    fn handle_event(&mut self, event: SwarmEvent<nomos_mix_network::Event>) {
+        match event {
+            SwarmEvent::Behaviour(nomos_mix_network::Event::FullyUnwrappedMessage(msg)) => {
+                tracing::debug!("Received fully unwrapped message: {msg:?}");
+                self.events_tx
+                    .send(Libp2pNetworkBackendEvent::FullyMixedMessage(msg))
+                    .unwrap();
+            }
+            SwarmEvent::Behaviour(nomos_mix_network::Event::Error(e)) => {
+                tracing::error!("Received error from mix network: {e:?}");
+            }
+            _ => {
+                tracing::debug!("Received event from mix network: {event:?}");
+            }
+        }
+    }
+}
+
+fn extract_peer_id(multiaddr: &Multiaddr) -> Option<PeerId> {
+    for protocol in multiaddr.iter() {
+        if let Protocol::P2p(peer_id) = protocol {
+            return Some(peer_id);
+        }
+    }
+    None
+}
```

Comment:
Nitpick, its totally fine like this, but you can use [`find](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find)` or [`find_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find_map) with iterators:

```suggestion
fn extract_peer_id(multiaddr: &Multiaddr) -> Option<PeerId> {
    multiaddr
        .iter()
        .find_map(|protocol| if let Some(peer_id) = protocol {Some(peer_id)} else { None })
}
```

Reply:
Thank you 3a6de0887c6f116db0d08377cf0af83f7d673562